Use JsonConverterAttribute in SerializeProperties

kevinoid opened the pull request:
If a property is declared with [`JsonConverterAttribute`](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_JsonConverterAttribute.htm) and has a non-`null`
value, use the converter to write the value in
[`JsonConverterHelper.SerializeProperties`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.rest.serialization.jsonconverterhelper.serializeproperties), to match Json.NET behavior.

Fixes: #12490

kevinoid committed:
Use JsonConverterAttribute for poly (de)serialize

Update Polymorphic(De)serializationJsonConverter to behave more like
Json.NET by using a property converter declared with
JsonConverterAttribute for writing non-null values if it can write, and
reading if it can read.

Fixes: #12490

Changes in v2:
- Check JsonConverter.CanRead/Write before calling.
- Use/Test JsonConverterAttribute when reading.
- Use UnixTimeJsonConverter for testing, since reading succeeds without
  JsonConverterAttribute for other DateTime formats.

Signed-off-by: Kevin Locke <kevin@kevinlocke.name>

azuresdkci commented:
Can one of the admins verify this patch?

jsquire reviewed (approved):
Hi @kevinoid.   Thank you very much for your contribution!   The changes look good to me.   I'm going to defer to @pakrym and @MiYanni for the final approval and merge.

kevinoid commented:
My pleasure.  Thanks @jsquire!

isra-fel reviewed (approved):
Looks good to me! Thanks for you contribution!

pakrym commented:
Do we need a similar change on the deserialization side?

kevinoid commented:
> Do we need a similar change on the deserialization side?

Yes indeed.  Good catch!

I've updated the PR to use `JsonConverterAttribute` for deserialization as well (and also check `CanRead`/`CanWrite` before calling the converter to match Json.NET).

pakrym commented:
LGTM

kevinoid commented:
I just noticed I had forgotten to push the `CanRead`/`CanWrite` check.  Sorry about that.  Now fixed.

Thanks all for reviewing it!

pakrym commented:
Can we add a test for the check?

kevinoid commented:
> Can we add a test for the check?

I was afraid you'd ask for that.  Ha.  Will do.

kevinoid committed:
Test poly (de)serialization CanRead/CanWrite

Ensure that .ReadJson() is called iff .CanRead and .WriteJson() is
called iff .CanWrite.

Signed-off-by: Kevin Locke <kevin@kevinlocke.name>

kevinoid commented:
I just added a commit with `CanRead`/`CanWrite` tests.  See what you think.

MiYanni commented on sdk/mgmtcommon/ClientRuntime/ClientRuntime/Serialization/JsonConverterHelper.cs:
```diff
@@ -77,7 +77,17 @@ public static void SerializeProperties(JsonWriter writer, object value, JsonSeri
+
+                    if (memberValue != null
+                        && property.Converter != null
+                        && property.Converter.CanWrite)
```
Can be `memberValue != null && property.Converter?.CanWrite == true`.

MiYanni commented on sdk/mgmtcommon/ClientRuntime/ClientRuntime/Serialization/PolymorphicDeserializeJsonConverter.cs:
```diff
@@ -103,7 +103,21 @@ private JToken SelectTokenCaseInsensitive(JObject obj, string path)
-                        var propertyValue = property.ToObject(expectedProperty.PropertyType, serializer);
+                        object propertyValue;
+                        if (expectedProperty.Converter != null
+                            && expectedProperty.Converter.CanRead)
```
Can be `expected.Converter?.CanRead == true`
